### ISSUE TITLE
Google Colab - Event Loop error on import autogluon

### ISSUE THREAD

**Issue by nathan-wood** [dask/distributed#3330]
Initially posted under autogluon repo, seems it might be a dask distributed error:
https://github.com/awslabs/autogluon/issues/162#issuecomment-566728430

To recreate the issue, here is a colab notebook for reference:
https://drive.google.com/open?id=1c14OY7zNP64xew1jlNpFGxb3D9No4u95

After installation, the first execution of the import throws an error.
Simply executing a 2nd time loads the library with no error.

---------------------
RuntimeError Traceback (most recent call last)
in ()
----> 1 import autogluon as ag

22 frames
/usr/lib/python3.6/asyncio/events.py in get_event_loop(self)
600 if self._local._loop is None:
601 raise RuntimeError('There is no current event loop in thread %r.'
--> 602 % threading.current_thread().name)
603 return self._local._loop
604

RuntimeError: There is no current event loop in thread 'IO loop'.

**Comment by mrocklin** [comment]
Hi @nathan-wood .  Thanks for raising the issue.  Unfortunately I'm not personally familiar with the `autogluon` library.  Are you able to reproduce this error using only Dask?  I'm not sure how else I can help without learning a lot about what autogluon is doing internally.

**Comment by nathan-wood** [comment]
Thanks @mrocklin. I'm not very familiar either, I'll dig a bit and see if I can isolate/reproduce with dask only

**Comment by quasiben** [comment]
@zhanghang1989 I don't think your error is related to this PR.  Still you could try doing:

> Client(process=False, dashboard_address=":8789")

if you continue experiencing issues please open another issue

**Comment by zhanghang1989** [comment]
> , dashboard_address=":8789"

The error is  related to this issue. I am main developer for autogluon toolkit and trying to debug the issue with Colab. AutoGluon toolkit highly depends on dask.distributed.

after trying your suggested command on Colab:
```
import distributed
remote = distributed.Client(processes=False, dashboard_address=":8789")
```
I got the error:
```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-1-4bb9bba26a5d> in <module>()
      1 import distributed
----> 2 remote = distributed.Client(processes=False, dashboard_address=":8789")

14 frames
/usr/local/lib/python3.6/dist-packages/distributed/client.py in __init__(self, address, loop, timeout, set_as_default, scheduler_file, security, asynchronous, name, heartbeat_interval, serializers, deserializers, extensions, direct_to_workers, **kwargs)
    725             ext(self)
    726 
--> 727         self.start(timeout=timeout)
    728         Client._instances.add(self)
    729 

/usr/local/lib/python3.6/dist-packages/distributed/client.py in start(self, **kwargs)
    890             self._started = asyncio.ensure_future(self._start(**kwargs))
    891         else:
--> 892             sync(self.loop, self._start, **kwargs)
    893 
    894     def __await__(self):

/usr/local/lib/python3.6/dist-packages/distributed/utils.py in sync(loop, func, callback_timeout, *args, **kwargs)
    332     if error[0]:
    333         typ, exc, tb = error[0]
--> 334         raise exc.with_traceback(tb)
    335     else:
    336         return result[0]

/usr/local/lib/python3.6/dist-packages/distributed/utils.py in f()
    316             if callback_timeout is not None:
    317                 future = gen.with_timeout(timedelta(seconds=callback_timeout), future)
--> 318             result[0] = yield future
    319         except Exception as exc:
    320             error[0] = sys.exc_info()

/usr/local/lib/python3.6/dist-packages/tornado/gen.py in run(self)
   1053 
   1054                     try:
-> 1055                         value = future.result()
   1056                     except Exception:
   1057                         self.had_exception = True

/usr/local/lib/python3.6/dist-packages/tornado/concurrent.py in result(self, timeout)
    236         if self._exc_info is not None:
    237             try:
--> 238                 raise_exc_info(self._exc_info)
    239             finally:
    240                 self = None

/usr/local/lib/python3.6/dist-packages/tornado/util.py in raise_exc_info(exc_info)

/usr/local/lib/python3.6/dist-packages/tornado/gen.py in wrapper(*args, **kwargs)
    305                 try:
    306                     orig_stack_contexts = stack_context._state.contexts
--> 307                     yielded = next(result)
    308                     if stack_context._state.contexts is not orig_stack_contexts:
    309                         yielded = TracebackFuture()

/usr/local/lib/python3.6/dist-packages/tornado/gen.py in _wrap_awaitable(x)

/usr/local/lib/python3.6/dist-packages/distributed/client.py in _start(self, timeout, **kwargs)
    956                     loop=self.loop,
    957                     asynchronous=self._asynchronous,
--> 958                     **self._startup_kwargs
    959                 )
    960             except (OSError, socket.error) as e:

/usr/local/lib/python3.6/dist-packages/distributed/deploy/spec.py in _()
    362         async def _():
    363             if self.status == "created":
--> 364                 await self._start()
    365             await self.scheduler
    366             await self._correct_state()

/usr/local/lib/python3.6/dist-packages/distributed/deploy/spec.py in _start(self)
    265             raise ValueError("Cluster is closed")
    266 
--> 267         self._lock = asyncio.Lock()
    268 
    269         if self.scheduler_spec is None:

/usr/lib/python3.6/asyncio/locks.py in __init__(self, loop)
    147             self._loop = loop
    148         else:
--> 149             self._loop = events.get_event_loop()
    150 
    151     def __repr__(self):

/usr/lib/python3.6/asyncio/events.py in get_event_loop()
    692     if current_loop is not None:
    693         return current_loop
--> 694     return get_event_loop_policy().get_event_loop()
    695 
    696 

/usr/lib/python3.6/asyncio/events.py in get_event_loop(self)
    600         if self._local._loop is None:
    601             raise RuntimeError('There is no current event loop in thread %r.'
--> 602                                % threading.current_thread().name)
    603         return self._local._loop
    604 

RuntimeError: There is no current event loop in thread 'IO loop'.
```

``!pip show distributed`` I got:
```
Name: distributed
Version: 2.9.0
Summary: Distributed scheduler for Dask
Home-page: https://distributed.dask.org
Author: None
Author-email: None
License: BSD
Location: /usr/local/lib/python3.6/dist-packages
Requires: msgpack, sortedcontainers, dask, zict, tblib, toolz, cloudpickle, click, tornado, pyyaml, psutil
Required-by: featuretools, autogluon
```

**Comment by quasiben** [comment]
It appears that the standard env on google colab requires tornado ~4.5:
> ERROR: google-colab 1.0.0 has requirement tornado~=4.5.0, but you'll have tornado 5.1 which is incompatible.

But latest distributed uses tornado >5.  Can I ask you try out what the rapids team does:

- [colab env](https://colab.research.google.com/drive/1rY7Ln6rEE1pOlfSHCYOVaqt8OvDO35J0#forceEdit=true&offline=true&sandboxMode=true)

- [rapids-setup](https://raw.githubusercontent.com/rapidsai/notebooks-contrib/890b04ed8687da6e3a100c81f449ff6f7b559956/utils/rapids-colab.sh)

**Comment by quasiben** [comment]
You may also be interested in these issues related to colab and dask:

- https://github.com/googlecolab/colabtools/issues/569
- https://github.com/dask/dask/issues/4843

**Comment by mrocklin** [comment]
Thank you @zhanghang1989 for investigating this and finding the problem.  I appreciate it.  It is always a challenge to figure out why a different software package doesn't work.

When I run the following:

```python
import distributed
remote = distributed.Client(processes=False)
```

I also get that warning, 

```python
/usr/local/lib/python3.6/dist-packages/distributed/bokeh/core.py:57: UserWarning: 
Failed to start diagnostics server on port 8787. [Errno 99] Cannot assign requested address
  warnings.warn('\n' + msg)
/usr/local/lib/python3.6/dist-packages/distributed/deploy/local.py:197: UserWarning: 
Could not launch service 'bokeh' on port 8787. Got the following message:

[Errno 99] Cannot assign requested address
  self.scheduler.start(scheduler_address)
```

but things do still work.  The simplest resolution to this might be to turn off the dashboard

```python
remote = distributed.Client(processes=False, dashboard_address=None)
```

I tried pip installing dask and distributed in colab, but then I found that colab itself crashed, perhaps due to the tornado issue mentioned above.  Have you run into this before @zhanghang1989 ?  It seems that you have worked through it.  I must admit that I do not often use Google Colab.

**Comment by zhanghang1989** [comment]
Thanks for the help @quasiben @mrocklin !

I do not often use Google Colab either. I mainly use AWS :)
I haven't worked through it before. Thanks for pointing out the potential issue of tornado! I will take a look.

**Comment by mrocklin** [comment]
> I do not often use Google Colab either. I mainly use AWS :)

I was surprised by this when I first read the issue :)

I'm also curious to learn more about how AutoGluon uses Dask.  I quickly searched through your docs but didn't find anything.  Is there somewhere that I can learn more?

**Comment by zhanghang1989** [comment]
AutoGluon schedulers are using dask distributed (Client, Scheduler, Queue). Mainly added GPU management and early stopping. https://github.com/awslabs/autogluon/blob/master/autogluon/scheduler/remote/remote.py#L48

**Comment by Innixma** [comment]
Root cause identified in https://github.com/awslabs/autogluon/issues/163. This has to do with differing tornado versions but we have found a work-around. I don't believe this is related to dask in particular. Please feel free to resolve, thanks!

**Comment by mrocklin** [comment]
Thanks for checking in @Innixma .  Closing.

Thanks also to @nathan-wood for originally reporting and @zhanghang1989 for helping out
